File: tests/integration_tests/orderbook/test_orderbook.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2023 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
import pytest

from nautilus_trader.model.enums import BookType
from nautilus_trader.model.objects import Price
from nautilus_trader.model.orderbook import OrderBook
from nautilus_trader.test_kit.stubs.data import TestDataStubs
from nautilus_trader.test_kit.stubs.identifiers import TestIdStubs
from tests import TEST_DATA_DIR


class TestOrderBook:
    def test_l1_orderbook(self):
        book = OrderBook(
            instrument_id=TestIdStubs.audusd_id(),
            book_type=BookType.L1_MBP,
        )
        i = 0
        for i, m in enumerate(TestDataStubs.l1_feed()):
            if m["op"] == "update":
                book.update(order=m["order"], ts_event=0)
            else:
                raise KeyError
            book.check_integrity()
        assert i == 1999

    def test_l2_feed(self):
        filename = TEST_DATA_DIR / "bitmex" / "L2_feed.json"

        book = OrderBook(
            instrument_id=TestIdStubs.audusd_id(),
            book_type=BookType.L2_MBP,
        )

        # Duplicate delete messages
        skip = [
            (12152, "378a3caf-0262-4d8b-95b6-8df65312b9f3"),
            (28646, "8101452c-8a80-4ca9-b0d9-c472691cec28"),
            (68431, "8913f4bf-cc49-4e23-b05d-5eeed948a454"),
        ]
        i = 0
        for i, m in enumerate(TestDataStubs.l2_feed(filename)):
            if not m or m["op"] == "trade":
                pass
            elif (i, m["order"].order_id) in skip:
                continue
            elif m["op"] == "update":
                book.update(order=m["order"], ts_event=0)
            elif m["op"] == "delete":
                book.delete(order=m["order"], ts_event=0)
            book.check_integrity()
        assert i == 68462

    @pytest.mark.skip("segfault on check_integrity")
    def test_l3_feed(self):
        filename = TEST_DATA_DIR / "bitmex" / "L3_feed.json"

        book = OrderBook(
            instrument_id=TestIdStubs.audusd_id(),
            book_type=BookType.L3_MBO,
        )

        # Updates that cause the book to fail integrity checks will be deleted
        # immediately, however we may also delete later.
        skip_deletes = []
        i = 0
        for i, m in enumerate(TestDataStubs.l3_feed(filename)):
            if m["op"] == "update":
                book.update(order=m["order"], ts_event=0)
                try:
                    book.check_integrity()
                except RuntimeError:  # BookIntegrityError was removed
                    book.delete(order=m["order"], ts_event=0)
                    skip_deletes.append(m["order"].order_id)
            elif m["op"] == "delete" and m["order"].order_id not in skip_deletes:
                book.delete(order=m["order"], ts_event=0)
            book.check_integrity()
        assert i == 100_047
        assert book.best_ask_level().price == 61405.27923706
        assert book.best_ask_level().volume() == 0.12227
        assert book.best_bid_level().price == Price.from_int(61_391)
        assert book.best_bid_level().volume() == 1

File: crates/adapters/coinbase/examples/test_orderbook.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Example demonstrating order book management with Coinbase Advanced Trade API.

use std::env;

use anyhow::Result;
use nautilus_coinbase::{
    orderbook::OrderBook,
    websocket::{client::CoinbaseWebSocketClient, types::WebSocketMessage, Channel},
};
use tracing::{info, warn};
use tracing_subscriber;

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .init();

    // Initialize Rustls crypto provider (REQUIRED for WebSocket TLS)
    let _ = rustls::crypto::aws_lc_rs::default_provider().install_default();

    info!("🚀 Coinbase Order Book Manager Example");
    info!("{}", "=".repeat(80));

    // Get API credentials from environment
    let api_key = env::var("COINBASE_API_KEY").expect("COINBASE_API_KEY must be set");
    let api_secret = env::var("COINBASE_API_SECRET").expect("COINBASE_API_SECRET must be set");

    // Create WebSocket client
    let client = CoinbaseWebSocketClient::new_market_data(api_key, api_secret);

    // Connect to WebSocket
    info!("🔌 Connecting to Coinbase WebSocket...");
    client.connect().await?;
    info!("✅ Connected!");

    // Subscribe to Level2 channel for BTC-USD
    let product_id = "BTC-USD";
    info!("📡 Subscribing to Level2 channel for {}...", product_id);
    client
        .subscribe(vec![product_id.to_string()], Channel::Level2)
        .await?;
    info!("✅ Subscribed!");

    // Create order book manager
    let mut orderbook = OrderBook::new(product_id.to_string());
    info!("📖 Order book manager created for {}", product_id);

    // Process messages
    info!("\n📨 Processing Level2 updates...\n");

    let mut snapshot_received = false;
    let mut update_count = 0;
    let max_updates = 20;

    loop {
        match client.receive_message().await? {
            Some(msg) => {
                match serde_json::from_str::<WebSocketMessage>(&msg) {
                    Ok(ws_msg) => {
                        if let WebSocketMessage::Level2 { events } = ws_msg {
                            for event in events {
                                // Process the event
                                orderbook.process_event(&event)?;

                                if event.event_type == "snapshot" {
                                    snapshot_received = true;
                                    info!("📸 Snapshot received!");
                                    info!("   Bids: {} levels", orderbook.bids.depth());
                                    info!("   Asks: {} levels", orderbook.asks.depth());

                                    // Display top 5 levels
                                    info!("\n   Top 5 Bids:");
                                    for (i, level) in orderbook.bids.top_levels(5).iter().enumerate() {
                                        info!(
                                            "   {}. ${} - {}",
                                            i + 1,
                                            level.price,
                                            level.size
                                        );
                                    }

                                    info!("\n   Top 5 Asks:");
                                    for (i, level) in orderbook.asks.top_levels(5).iter().enumerate() {
                                        info!(
                                            "   {}. ${} - {}",
                                            i + 1,
                                            level.price,
                                            level.size
                                        );
                                    }

                                    // Display market metrics
                                    if let Some(best_bid) = orderbook.best_bid() {
                                        info!("\n   Best Bid: ${}", best_bid);
                                    }
                                    if let Some(best_ask) = orderbook.best_ask() {
                                        info!("   Best Ask: ${}", best_ask);
                                    }
                                    if let Some(mid) = orderbook.mid_price() {
                                        info!("   Mid Price: ${}", mid);
                                    }
                                    if let Some(spread) = orderbook.spread() {
                                        info!("   Spread: ${}", spread);
                                    }
                                    if let Some(spread_bps) = orderbook.spread_bps() {
                                        info!("   Spread (bps): {:.2}", spread_bps);
                                    }
                                } else if snapshot_received {
                                    update_count += 1;

                                    if update_count <= 5 || update_count % 5 == 0 {
                                        info!(
                                            "📊 Update #{}: {} changes | Bids: {} | Asks: {} | Mid: ${} | Spread: {:.2} bps",
                                            update_count,
                                            event.updates.len(),
                                            orderbook.bids.depth(),
                                            orderbook.asks.depth(),
                                            orderbook.mid_price().unwrap_or_default(),
                                            orderbook.spread_bps().unwrap_or_default()
                                        );
                                    }

                                    if update_count >= max_updates {
                                        info!("\n✅ Processed {} updates. Stopping...", update_count);
                                        break;
                                    }
                                }
                            }
                        }
                    }
                    Err(e) => {
                        warn!("Failed to parse WebSocket message: {}", e);
                    }
                }
            }
            None => {
                info!("WebSocket connection closed");
                break;
            }
        }
    }

    // Final statistics
    info!("\n{}", "=".repeat(80));
    info!("📊 Final Order Book Statistics");
    info!("{}", "=".repeat(80));
    info!("Product: {}", orderbook.product_id);
    info!("Bid Levels: {}", orderbook.bids.depth());
    info!("Ask Levels: {}", orderbook.asks.depth());
    if let Some(best_bid) = orderbook.best_bid() {
        info!("Best Bid: ${}", best_bid);
    }
    if let Some(best_ask) = orderbook.best_ask() {
        info!("Best Ask: ${}", best_ask);
    }
    if let Some(mid) = orderbook.mid_price() {
        info!("Mid Price: ${}", mid);
    }
    if let Some(spread) = orderbook.spread() {
        info!("Spread: ${}", spread);
    }
    if let Some(spread_bps) = orderbook.spread_bps() {
        info!("Spread (bps): {:.2}", spread_bps);
    }

    // Disconnect
    info!("\n🔌 Disconnecting...");
    client.disconnect().await?;
    info!("✅ Disconnected!");

    info!("\n🎉 Order book example completed successfully!");

    Ok(())
}

